File: 22/Problem22.py
class Solution:
    def generateParenthesis(self, n):
        result = []
        self.backtracking(result, 0, 0, n, "")
        return result

    def backtracking(self, result, open, closed, n, current):
        if open == n and closed == n:
            result.append(current)
            return

        if open < n:
            self.backtracking(result, open + 1, closed, n, current + "(")

        if closed < open:
            self.backtracking(result, open, closed + 1, n, current + ")")

if __name__ == "__main__":
    solution = Solution()
    n = 3
    combinations = solution.generateParenthesis(n)

    print(f"All combinations of well-formed parentheses for n = {n} are:")
    for combination in combinations:
        print(combination)

File: LeetCode/Backtracking/22/Problem22.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<string> generateParenthesis(int n) {
        // only add left parenthesis if open < n
        // only add right parenthesis if closed < open
        // valid if open == closed == n
        vector<string> result;
        backtracking(result, 0, 0, n, "");
        return result;
    }

    void backtracking(vector<string> &result, int open, int closed, int n, string current) {
        if(open == n && closed == n) {
            result.push_back(current);
            return;
        }
        if(open < n) 
            backtracking(result, open+1, closed, n, current + '(');
        if(closed < open)
            backtracking(result, open, closed+1, n, current + ')');
    }
};

int main() {
    Solution solution;
    int n = 3; 
    vector<string> combinations = solution.generateParenthesis(n);

    cout << "All combinations of well-formed parentheses for n = " << n << " are:\n";
    for (const string& combination : combinations) {
        cout << combination << endl;
    }

    return 0;
}